monitoring/google/cloud/monitoring/query.py

```diff
@@ -0,0 +1,604 @@
+        if self._start_time:
+          params['interval'].start_time.FromDatetime(self._start_time)
+
+        params['aggregation'] = types.Aggregation(
+            per_series_aligner=self._per_series_aligner,
+            cross_series_reducer=self._cross_series_reducer,
+            group_by_fields=self._group_by_fields,
+            alignment_period={'seconds': self._alignment_period_seconds},
+        )
+
+        if headers_only:
+            params['view'] = enums.ListTimeSeriesRequest.TimeSeriesView.HEADERS
+        else:
+            params['view'] = enums.ListTimeSeriesRequest.TimeSeriesView.FULL
+
+        if page_size is not None:
+            params['page_size'] = page_size
+
+        return params
+
+    def as_dataframe(self, label=None, labels=None):
+        """Return all the selected time series as a :mod:`pandas` dataframe.
+
+        .. note::
+
+            Use of this method requires that you have :mod:`pandas` installed.
+
+        Examples::
+
+            # Generate a dataframe with a multi-level column header including
+            # the resource type and all available resource and metric labels.
+            # This can be useful for seeing what labels are available.
+            dataframe = query.as_dataframe()
+
+            # Generate a dataframe using a particular label for the column
+            # names.
+            dataframe = query.as_dataframe(label='instance_name')
+
+            # Generate a dataframe with a multi-level column header.
+            dataframe = query.as_dataframe(labels=['zone', 'instance_name'])
+
+            # Generate a dataframe with a multi-level column header, assuming
+            # the metric is issued by more than one type of resource.
+            dataframe = query.as_dataframe(
+                labels=['resource_type', 'instance_id'])
+
+        :type label: str
+        :param label:
+            (Optional) The label name to use for the dataframe header.
+            This can be the name of a resource label or metric label
+            (e.g., ``"instance_name"``), or the string ``"resource_type"``.
+
+        :type labels: list of strings, or None
+        :param labels: A list or tuple of label names to use for the dataframe
+            header. If more than one label name is provided, the resulting
+            dataframe will have a multi-level column header. Providing values
+            for both ``label`` and ``labels`` is an error.
+
+        :rtype: :class:`pandas.DataFrame`
+        :returns: A dataframe where each column represents one time series.
+        """
+        return _build_dataframe(self, label, labels)  # pragma: NO COVER
+
+    def copy(self):
+        """Copy the query object.
+
+        :rtype: :class:`Query`
+        :returns: The new query object.
+        """
+        # Using copy.deepcopy() would be appropriate, except that we want
+        # to copy self._client only as a reference.
+        new_query = copy.copy(self)
```

Comment:
Why not implement [`__deepcopy__`](https://docs.python.org/3/library/copy.html) and copy everything in the memo dictionary except `_client`? That is what your comment suggests you want.

Reply:
Thanks for the suggestion. Done.
PTAL, and let me know if it makes sense.